datastore/user_keys.go

```diff
@@ -32,10 +34,40 @@ func (DBUserKey) TableName() string {
 
 // StoreUserKey saves a user key to the database
 func (d *Datastore) StoreUserKey(key *DBUserKey) error {
-	// Count existing keys for this service/account pair, excluding the current key_name
+	return d.DB.Transaction(func(tx *gorm.DB) error {
+		updated, err := d.updateExistingKeyTx(tx, key)
+		if err != nil {
+			return err
+		}
+		if updated {
+			return nil
+		}
+		return d.insertNewKeyTx(tx, key)
+	})
+}
+
+// updateExistingKeyTx attempts to update an existing key atomically
+func (d *Datastore) updateExistingKeyTx(tx *gorm.DB, key *DBUserKey) (bool, error) {
+	result := tx.Model(&DBUserKey{}).
+		Where("account_id = ? AND service = ? AND key_name = ?",
+			key.AccountID, key.Service, key.KeyName).
+		Updates(map[string]interface{}{
+			"key_material":  key.KeyMaterial,
+			"updated_at":    key.UpdatedAt,
```

Comment:
nit: Perhaps we should replace `updated_at` with `created_at`. It may be a better fit now with the serial number present. Thoughts?

Reply:
I think that's a good idea, but I'll do that in a follow-up to keep this PR smaller.